File: src/lexer.py
import enum
import re

class TokenType(enum.Enum):
    # Directives
    EIGEN = "eigen"
    
    # Types
    QUBIT = "qubit"
    CBIT = "cbit"
    INT = "int"
    FLOAT = "float"
    STRING = "string"
    BOOL = "bool"
    ARRAY = "array"
    MAP = "map"
    
    # Keywords
    MODULE = "module"
    IMPORT = "import"
    QFUNC = "qfunc"
    FUNC = "func"
    STRUCT = "struct"
    ENUM = "enum"
    TRY = "try"
    CATCH = "catch"
    THROW = "throw"
    NOISE = "noise"
    DEPOLARIZING = "depolarizing"
    BITFLIP = "bitflip"
    LET = "let"
    IF = "if"
    FOR = "for"
    IN = "in"
    WHILE = "while"
    BREAK = "break"
    CONTINUE = "continue"
    MEASURE = "measure"
    RETURN = "return"
    
    # Built-ins / Utilities
    TRACE = "trace"
    PRINT = "print"
    ASSERT = "assert"
    
    # Constants
    PI = "PI"
    TAU = "TAU"
    E = "E"
    
    # Gates
    GATE_H = "H"
    GATE_X = "X"
    GATE_Y = "Y"
    GATE_Z = "Z"
    GATE_S = "S"
    GATE_T = "T"
    GATE_CNOT = "CNOT"
    GATE_CZ = "CZ"
    GATE_SWAP = "SWAP"
    GATE_RX = "RX"
    GATE_RY = "RY"
    GATE_RZ = "RZ"
    
    # Literals
    STRING_LIT = "STRING_LIT"
    TRUE = "true"
    FALSE = "false"
    NULL = "null"
    
    # General
    IDENTIFIER = "IDENTIFIER"
    INT_LIT = "INT_LIT"
    FLOAT_LIT = "FLOAT_LIT"
    
    # Operators & Delimiters
    LPAREN = "("
    RPAREN = ")"
    LBRACE = "{"
    RBRACE = "}"
    LBRACK = "["
    RBRACK = "]"
    COMMA = ","
    COLON = ":"
    DOT = "."
    ARROW = "->"
    EQUALS = "="
    EQ = "=="
    NE = "!="
    LT = "<"
    GT = ">"
    LE = "<="
    GE = ">="
    
    PLUS = "+"
    MINUS = "-"
    MUL = "*"
    DIV = "/"
    
    ADD_ASSIGN = "+="
    SUB_ASSIGN = "-="
    MUL_ASSIGN = "*="
    DIV_ASSIGN = "/="
    
    AND = "and"
    OR = "or"
    NOT = "not"
    
    EOF = "EOF"


class Token:
    def __init__(self, type_: TokenType, value: str, line: int, column: int):
        self.type = type_
        self.value = value
        self.line = line
        self.column = column

    def __repr__(self):
        return f"Token({self.type.name}, {repr(self.value)}, line={self.line}, col={self.column})"


class Lexer:
    def __init__(self, source: str):
        self.source = source
        self.length = len(source)
        self.pos = 0
        self.line = 1
        self.column = 1

    def error(self, msg: str):
        raise SyntaxError(f"Lexer Error at line {self.line}, col {self.column}: {msg}")

    def peek(self, offset: int = 0) -> str:
        if self.pos + offset >= self.length:
            return ""
        return self.source[self.pos + offset]

    def advance(self):
        if self.pos < self.length:
            char = self.source[self.pos]
            self.pos += 1
            if char == '\n':
                self.line += 1
                self.column = 1
            else:
                self.column += 1

    def tokenize(self) -> list[Token]:
        tokens = []
        
        KEYWORDS_MAP = {
            "eigen": TokenType.EIGEN,
            "qubit": TokenType.QUBIT,
            "cbit": TokenType.CBIT,
            "int": TokenType.INT,
            "float": TokenType.FLOAT,
            "string": TokenType.STRING,
            "bool": TokenType.BOOL,
            "array": TokenType.ARRAY,
            "map": TokenType.MAP,
            "module": TokenType.MODULE,
            "import": TokenType.IMPORT,
            "qfunc": TokenType.QFUNC,
            "func": TokenType.FUNC,
            "struct": TokenType.STRUCT,
            "enum": TokenType.ENUM,
            "try": TokenType.TRY,
            "catch": TokenType.CATCH,
            "throw": TokenType.THROW,
            "noise": TokenType.NOISE,
            "depolarizing": TokenType.DEPOLARIZING,
            "bitflip": TokenType.BITFLIP,
            "let": TokenType.LET,
            "if": TokenType.IF,
            "for": TokenType.FOR,
            "in": TokenType.IN,
            "while": TokenType.WHILE,
            "break": TokenType.BREAK,
            "continue": TokenType.CONTINUE,
            "measure": TokenType.MEASURE,
            "return": TokenType.RETURN,
            "trace": TokenType.TRACE,
            "print": TokenType.PRINT,
            "assert": TokenType.ASSERT,
            "PI": TokenType.PI,
            "TAU": TokenType.TAU,
            "E": TokenType.E,
            "H": TokenType.GATE_H,
            "X": TokenType.GATE_X,
            "Y": TokenType.GATE_Y,
            "Z": TokenType.GATE_Z,
            "S": TokenType.GATE_S,
            "T": TokenType.GATE_T,
            "CNOT": TokenType.GATE_CNOT,
            "CZ": TokenType.GATE_CZ,
            "SWAP": TokenType.GATE_SWAP,
            "RX": TokenType.GATE_RX,
            "RY": TokenType.GATE_RY,
            "RZ": TokenType.GATE_RZ,
            "true": TokenType.TRUE,
            "false": TokenType.FALSE,
            "null": TokenType.NULL,
            "and": TokenType.AND,
            "or": TokenType.OR,
            "not": TokenType.NOT,
        }

        while self.pos < self.length:
            char = self.peek()

            # Skip whitespace
            if char.isspace():
                self.advance()
                continue

            # Skip comments (both # and //)
            if char == '#' or (char == '/' and self.peek(1) == '/'):
                while self.pos < self.length and self.peek() != '\n':
                    self.advance()
                continue

            # Double-quoted string literals
            if char == '"':
                start_col = self.column
                self.advance()  # consume open quote
                string_val = ""
                while self.pos < self.length and self.peek() != '"':
                    if self.peek() == '\n':
                        self.error("Unterminated string literal")
                    # Handle escape sequences
                    if self.peek() == '\\':
                        self.advance()
                        if self.pos < self.length:
                            string_val += self.peek()
                            self.advance()
                    else:
                        string_val += self.peek()
                        self.advance()
                if self.peek() != '"':
                    self.error("Unterminated string literal")
                self.advance()  # consume close quote
                tokens.append(Token(TokenType.STRING_LIT, string_val, self.line, start_col))
                continue

            # Multi-character operators and symbols
            if char == '-' and self.peek(1) == '>':
                start_col = self.column
                self.advance()
                self.advance()
                tokens.append(Token(TokenType.ARROW, "->", self.line, start_col))
                continue

            if char == '=' and self.peek(1) == '=':
                start_col = self.column
                self.advance()
                self.advance()
                tokens.append(Token(TokenType.EQ, "==", self.line, start_col))
                continue

            if char == '!' and self.peek(1) == '=':
                start_col = self.column
                self.advance()
                self.advance()
                tokens.append(Token(TokenType.NE, "!=", self.line, start_col))
                continue

            if char == '<' and self.peek(1) == '=':
                start_col = self.column
                self.advance()
                self.advance()
                tokens.append(Token(TokenType.LE, "<=", self.line, start_col))
                continue

            if char == '>' and self.peek(1) == '=':
                start_col = self.column
                self.advance()
                self.advance()
                tokens.append(Token(TokenType.GE, ">=", self.line, start_col))
                continue

            if char == '+' and self.peek(1) == '=':
                start_col = self.column
                self.advance()
                self.advance()
                tokens.append(Token(TokenType.ADD_ASSIGN, "+=", self.line, start_col))
                continue

            if char == '-' and self.peek(1) == '=':
                start_col = self.column
                self.advance()
                self.advance()
                tokens.append(Token(TokenType.SUB_ASSIGN, "-=", self.line, start_col))
                continue

            if char == '*' and self.peek(1) == '=':
                start_col = self.column
                self.advance()
                self.advance()
                tokens.append(Token(TokenType.MUL_ASSIGN, "*=", self.line, start_col))
                continue

            if char == '/' and self.peek(1) == '=':
                start_col = self.column
                self.advance()
                self.advance()
                tokens.append(Token(TokenType.DIV_ASSIGN, "/=", self.line, start_col))
                continue

            # Single character operators & delimiters
            char_tokens = {
                '(': TokenType.LPAREN,
                ')': TokenType.RPAREN,
                '{': TokenType.LBRACE,
                '}': TokenType.RBRACE,
                '[': TokenType.LBRACK,
                ']': TokenType.RBRACK,
                ',': TokenType.COMMA,
                ':': TokenType.COLON,
                '.': TokenType.DOT,
                '=': TokenType.EQUALS,
                '+': TokenType.PLUS,
                '-': TokenType.MINUS,
                '*': TokenType.MUL,
                '/': TokenType.DIV,
                '<': TokenType.LT,
                '>': TokenType.GT,
            }
            if char in char_tokens:
                tokens.append(Token(char_tokens[char], char, self.line, self.column))
                self.advance()
                continue

            # Numbers (integers or floats)
            if char.isdigit():
                start_col = self.column
                num_str = ""
                while self.pos < self.length and self.peek().isdigit():
                    num_str += self.peek()
                    self.advance()
                
                # Check for decimal point
                if self.peek() == '.' and self.peek(1).isdigit():
                    num_str += '.'
                    self.advance()
                    while self.pos < self.length and self.peek().isdigit():
                        num_str += self.peek()
                        self.advance()
                    tokens.append(Token(TokenType.FLOAT_LIT, num_str, self.line, start_col))
                else:
                    tokens.append(Token(TokenType.INT_LIT, num_str, self.line, start_col))
                continue

            # Identifiers and keywords (no dots allowed anymore, dots are separate tokens)
            if char.isalpha() or char == '_':
                start_col = self.column
                ident_str = ""
                while self.pos < self.length and (self.peek().isalnum() or self.peek() == '_'):
                    ident_str += self.peek()
                    self.advance()

                if ident_str in KEYWORDS_MAP:
                    tokens.append(Token(KEYWORDS_MAP[ident_str], ident_str, self.line, start_col))
                else:
                    tokens.append(Token(TokenType.IDENTIFIER, ident_str, self.line, start_col))
                continue

            self.error(f"Unexpected character: {repr(char)}")

        tokens.append(Token(TokenType.EOF, "", self.line, self.column))
        return tokens

File: native/rust/src/frontend/lexer.rs
use serde::Serialize;

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize)]
pub enum TokenType {
    // Directives
    Eigen,
    // Types
    Qubit, Cbit, Int, Float, StringType, Bool, Array, Map,
    // Keywords
    Module, Import, Qfunc, Func, Struct, Enum, Try, Catch, Throw, Noise, Depolarizing, Bitflip, Let, If, For, In, While, Break, Continue, Measure, Return, Parallel, Task,
    // Built-ins / Utilities
    Trace, Print, Assert,
    // Constants
    Pi, Tau, E,
    // Gates
    GateH, GateX, GateY, GateZ, GateS, GateT, GateCnot, GateCz, GateSwap, GateRx, GateRy, GateRz,
    // Literals
    StringLit, True, False, Null,
    // General
    Identifier, IntLit, FloatLit,
    // Operators & Delimiters
    Lparen, Rparen, Lbrace, Rbrace, Lbrack, Rbrack, Comma, Colon, Dot, Arrow, Equals, Eq, Ne, Lt, Gt, Le, Ge,
    Plus, Minus, Mul, Div,
    AddAssign, SubAssign, MulAssign, DivAssign,
    And, Or, Not,
    Eof,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize)]
pub struct Token<'a> {
    pub token_type: TokenType,
    pub value: &'a str,
    pub line: usize,
    pub column: usize,
}

pub struct Lexer<'a> {
    source: &'a str,
    pos: usize,
    line: usize,
    column: usize,
}

impl<'a> Lexer<'a> {
    pub fn new(source: &'a str) -> Self {
        Self {
            source,
            pos: 0,
            line: 1,
            column: 1,
        }
    }

    fn error(&self, msg: &str) -> String {
        format!("Lexer Error at line {}, col {}: {}", self.line, self.column, msg)
    }

    fn peek(&self, offset: usize) -> Option<char> {
        self.source[self.pos..].chars().nth(offset)
    }

    fn advance(&mut self) {
        if self.pos < self.source.len() {
            let ch = self.source[self.pos..].chars().next().unwrap();
            self.pos += ch.len_utf8();
            if ch == '\n' {
                self.line += 1;
                self.column = 1;
            } else {
                self.column += 1;
            }
        }
    }

    pub fn tokenize(&mut self) -> Result<Vec<Token<'a>>, String> {
        let mut tokens = Vec::new();

        while self.pos < self.source.len() {
            let current_char = self.source[self.pos..].chars().next().unwrap();

            // Skip whitespace
            if current_char.is_whitespace() {
                if current_char == '\n' {
                    self.line += 1;
                    self.column = 1;
                    self.pos += 1;
                } else {
                    let start_pos = self.pos;
                    while self.pos < self.source.len() {
                        let ch = self.source[self.pos..].chars().next().unwrap();
                        if ch.is_whitespace() && ch != '\n' {
                            self.pos += ch.len_utf8();
                        } else {
                            break;
                        }
                    }
                    self.column += self.pos - start_pos;
                }
                continue;
            }

            // Skip comments (# and //)
            if current_char == '#' || (current_char == '/' && self.peek(1) == Some('/')) {
                while self.pos < self.source.len() {
                    let ch = self.source[self.pos..].chars().next().unwrap();
                    if ch == '\n' {
                        break;
                    }
                    self.pos += ch.len_utf8();
                }
                // We do NOT consume the '\n' because whitespace check will handle it to increment line count.
                continue;
            }

            // Double-quoted string literals
            if current_char == '"' {
                let start_col = self.column;
                let start_line = self.line;
                self.advance(); // consume open quote
                let start_pos = self.pos;

                while self.pos < self.source.len() {
                    let ch = self.source[self.pos..].chars().next().unwrap();
                    if ch == '"' {
                        break;
                    }
                    if ch == '\n' {
                        return Err(format!("Lexer Error at line {}, col {}: Unterminated string literal", start_line, start_col));
                    }
                    if ch == '\\' {
                        self.advance(); // consume '\\'
                        if self.pos < self.source.len() {
                            self.advance(); // consume escaped char
                        }
                    } else {
                        self.advance();
                    }
                }

                if self.pos >= self.source.len() || self.source[self.pos..].chars().next() != Some('"') {
                    return Err(format!("Lexer Error at line {}, col {}: Unterminated string literal", start_line, start_col));
                }

                let literal_val = &self.source[start_pos..self.pos];
                self.advance(); // consume close quote
                tokens.push(Token {
                    token_type: TokenType::StringLit,
                    value: literal_val,
                    line: start_line,
                    column: start_col,
                });
                continue;
            }

            // Multi-character operators and symbols
            if current_char == '-' && self.peek(1) == Some('>') {
                let start_col = self.column;
                self.advance();
                self.advance();
                tokens.push(Token { token_type: TokenType::Arrow, value: "->", line: self.line, column: start_col });
                continue;
            }

            if current_char == '=' && self.peek(1) == Some('=') {
                let start_col = self.column;
                self.advance();
                self.advance();
                tokens.push(Token { token_type: TokenType::Eq, value: "==", line: self.line, column: start_col });
                continue;
            }

            if current_char == '!' && self.peek(1) == Some('=') {
                let start_col = self.column;
                self.advance();
                self.advance();
                tokens.push(Token { token_type: TokenType::Ne, value: "!=", line: self.line, column: start_col });
                continue;
            }

            if current_char == '<' && self.peek(1) == Some('=') {
                let start_col = self.column;
                self.advance();
                self.advance();
                tokens.push(Token { token_type: TokenType::Le, value: "<=", line: self.line, column: start_col });
                continue;
            }

            if current_char == '>' && self.peek(1) == Some('=') {
                let start_col = self.column;
                self.advance();
                self.advance();
                tokens.push(Token { token_type: TokenType::Ge, value: ">=", line: self.line, column: start_col });
                continue;
            }

            if current_char == '+' && self.peek(1) == Some('=') {
                let start_col = self.column;
                self.advance();
                self.advance();
                tokens.push(Token { token_type: TokenType::AddAssign, value: "+=", line: self.line, column: start_col });
                continue;
            }

            if current_char == '-' && self.peek(1) == Some('=') {
                let start_col = self.column;
                self.advance();
                self.advance();
                tokens.push(Token { token_type: TokenType::SubAssign, value: "-=", line: self.line, column: start_col });
                continue;
            }

            if current_char == '*' && self.peek(1) == Some('=') {
                let start_col = self.column;
                self.advance();
                self.advance();
                tokens.push(Token { token_type: TokenType::MulAssign, value: "*=", line: self.line, column: start_col });
                continue;
            }

            if current_char == '/' && self.peek(1) == Some('=') {
                let start_col = self.column;
                self.advance();
                self.advance();
                tokens.push(Token { token_type: TokenType::DivAssign, value: "/=", line: self.line, column: start_col });
                continue;
            }

            // Single character tokens
            let single_char_token = match current_char {
                '(' => Some(TokenType::Lparen),
                ')' => Some(TokenType::Rparen),
                '{' => Some(TokenType::Lbrace),
                '}' => Some(TokenType::Rbrace),
                '[' => Some(TokenType::Lbrack),
                ']' => Some(TokenType::Rbrack),
                ',' => Some(TokenType::Comma),
                ':' => Some(TokenType::Colon),
                '.' => Some(TokenType::Dot),
                '=' => Some(TokenType::Equals),
                '+' => Some(TokenType::Plus),
                '-' => Some(TokenType::Minus),
                '*' => Some(TokenType::Mul),
                '/' => Some(TokenType::Div),
                '<' => Some(TokenType::Lt),
                '>' => Some(TokenType::Gt),
                _ => None,
            };

            if let Some(t_type) = single_char_token {
                let val = &self.source[self.pos..self.pos + 1];
                tokens.push(Token {
                    token_type: t_type,
                    value: val,
                    line: self.line,
                    column: self.column,
                });
                self.advance();
                continue;
            }

            // Numbers
            if current_char.is_ascii_digit() {
                let start_pos = self.pos;
                let start_col = self.column;
                while self.pos < self.source.len() && self.source[self.pos..].chars().next().unwrap().is_ascii_digit() {
                    self.advance();
                }

                // Check for decimal point
                if self.pos < self.source.len() && self.source[self.pos..].starts_with('.') {
                    if let Some(next_ch) = self.peek(1) {
                        if next_ch.is_ascii_digit() {
                            self.advance(); // consume '.'
                            while self.pos < self.source.len() && self.source[self.pos..].chars().next().unwrap().is_ascii_digit() {
                                self.advance();
                            }
                            let val = &self.source[start_pos..self.pos];
                            tokens.push(Token {
                                token_type: TokenType::FloatLit,
                                value: val,
                                line: self.line,
                                column: start_col,
                            });
                            continue;
                        }
                    }
                }

                let val = &self.source[start_pos..self.pos];
                tokens.push(Token {
                    token_type: TokenType::IntLit,
                    value: val,
                    line: self.line,
                    column: start_col,
                });
                continue;
            }

            // Identifiers / Keywords
            if current_char.is_alphabetic() || current_char == '_' {
                let start_pos = self.pos;
                let start_col = self.column;
                while self.pos < self.source.len() {
                    let ch = self.source[self.pos..].chars().next().unwrap();
                    if ch.is_alphanumeric() || ch == '_' {
                        self.advance();
                    } else {
                        break;
                    }
                }
                let val = &self.source[start_pos..self.pos];
                let t_type = match val {
                    "eigen" => TokenType::Eigen,
                    "qubit" => TokenType::Qubit,
                    "cbit" => TokenType::Cbit,
                    "int" => TokenType::Int,
                    "float" => TokenType::Float,
                    "string" => TokenType::StringType,
                    "bool" => TokenType::Bool,
                    "array" => TokenType::Array,
                    "map" => TokenType::Map,
                    "module" => TokenType::Module,
                    "import" => TokenType::Import,
                    "qfunc" => TokenType::Qfunc,
                    "func" => TokenType::Func,
                    "struct" => TokenType::Struct,
                    "enum" => TokenType::Enum,
                    "try" => TokenType::Try,
                    "catch" => TokenType::Catch,
                    "throw" => TokenType::Throw,
                    "noise" => TokenType::Noise,
                    "depolarizing" => TokenType::Depolarizing,
                    "bitflip" => TokenType::Bitflip,
                    "let" => TokenType::Let,
                    "if" => TokenType::If,
                    "for" => TokenType::For,
                    "in" => TokenType::In,
                    "while" => TokenType::While,
                    "break" => TokenType::Break,
                    "continue" => TokenType::Continue,
                    "measure" => TokenType::Measure,
                    "return" => TokenType::Return,
                    "parallel" => TokenType::Parallel,
                    "task" => TokenType::Task,
                    "trace" => TokenType::Trace,
                    "print" => TokenType::Print,
                    "assert" => TokenType::Assert,
                    "PI" => TokenType::Pi,
                    "TAU" => TokenType::Tau,
                    "E" => TokenType::E,
                    "H" => TokenType::GateH,
                    "X" => TokenType::GateX,
                    "Y" => TokenType::GateY,
                    "Z" => TokenType::GateZ,
                    "S" => TokenType::GateS,
                    "T" => TokenType::GateT,
                    "CNOT" => TokenType::GateCnot,
                    "CZ" => TokenType::GateCz,
                    "SWAP" => TokenType::GateSwap,
                    "RX" => TokenType::GateRx,
                    "RY" => TokenType::GateRy,
                    "RZ" => TokenType::GateRz,
                    "true" => TokenType::True,
                    "false" => TokenType::False,
                    "null" => TokenType::Null,
                    "and" => TokenType::And,
                    "or" => TokenType::Or,
                    "not" => TokenType::Not,
                    _ => TokenType::Identifier,
                };

                tokens.push(Token {
                    token_type: t_type,
                    value: val,
                    line: self.line,
                    column: start_col,
                });
                continue;
            }

            return Err(self.error(&format!("Unexpected character: {:?}", current_char)));
        }

        tokens.push(Token {
            token_type: TokenType::Eof,
            value: "",
            line: self.line,
            column: self.column,
        });

        Ok(tokens)
    }
}
